design/implicit-boot-mode.md

```diff
@@ -0,0 +1,197 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# implicit-boot-mode
+
+## Status
+
+provisional
+
+## Table of Contents
+
+<!--ts-->
+   * [Title](#title)
+      * [Status](#status)
+      * [Table of Contents](#table-of-contents)
+      * [Summary](#summary)
+      * [Motivation](#motivation)
+         * [Goals](#goals)
+         * [Non-Goals](#non-goals)
+      * [Proposal](#proposal)
+         * [User Stories [optional]](#user-stories-optional)
+            * [Story 1](#story-1)
+            * [Story 2](#story-2)
+         * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+         * [Risks and Mitigations](#risks-and-mitigations)
+      * [Design Details](#design-details)
+         * [Work Items](#work-items)
+         * [Dependencies](#dependencies)
+         * [Test Plan](#test-plan)
+         * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+         * [Version Skew Strategy](#version-skew-strategy)
+      * [Drawbacks [optional]](#drawbacks-optional)
+      * [Alternatives [optional]](#alternatives-optional)
+      * [References](#references)
+
+<!-- Added by: stack, at: 2019-02-15T11:41-05:00 -->
+
+<!--te-->
+
+## Summary
+
+The required boot mode for a host depends on several factors,
+including support in the hardware, whether IPv6 is being used, and
+whether virtual media is being used. We do not need to support all
+combinations of those features with all boot modes. Asking the user to
+give us the boot mode explicitly introduces an opportunity for them to
+specify an incorrect value. We have enough information about the host
```

Comment:
Just out of curiosity, if we _were_ to implement explicit config of ```boot_mode```, just the way it is in Ironic, it would be an addition to what this proposal talks about, yes?

Reply:
I think they begin to become mutually exclusive. I guess what I'm lacking is a clear definition of why bios/legacy mode is still required. I'd personally love to see the default change, and all the complexity here is really around trying to avoid some of the issues that are present with different BMC implementations. Hiding everything _does_ make it easier for us to define our supported configurations, and reduce the risk of being distracted by someone trying to essentially deploy/operate systems in modes that we wouldn't want them to operate in under normal circumstances.

Reply:
While the point of this proposal is to try not to add a new API field, if we did I see a few ways we could use it:

1. We could require that the user always provide it.
2. We could have a single default and let the API override that. 
3. We could keep the system described here, and let the API override that.

As to why we need to keep IPMI, I tried to explain that briefly down around line 108. The tl;dr version of this proposal is "always use UEFI, except when using IPMI".

Reply:
I think 3. sounds good. And because of 3. and given Julia's comment that both of these are mutually exclusive, this work, if not anything else, becomes an augmentation to the future (if any) additions to the API (whenever/however they are needed). 1. and 3. would be the way to go, in my opinion.
I agree with the pain IPMI provides. And I think it sounds more like an Ironic community wide discusison getting everybody on board, before we can move away from the legacy/IPMI setup. Ergo, it's safe to ignore that/keep the way it is in this proposal.